backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+				continue;
+			}
+
+			final IQueueProcessor queueProcessor = queueProcessorOptional.get();
+
+			final Optional<I_C_Queue_WorkPackage> workPackageWithOwnCtx = getWorkPackageWithOwnContext(queueProcessor, workPackage, ctx);
+
+			if (!workPackageWithOwnCtx.isPresent())
+			{
+				logger.warn("*** processWorkPackages: Couldn't assign a new Ctx to C_Queue_WorkPackage: {}", workPackage.getC_Queue_WorkPackage_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final boolean successfullyHandled = handleWorkPackageProcessing(queueProcessor, workPackageWithOwnCtx.get());
+
+			handledAllWorkPackages = handledAllWorkPackages && successfullyHandled;
+		}
+
+		return handledAllWorkPackages;
+	}
+
+	@NonNull
+	private List<I_C_Queue_WorkPackage> pollAndLockWorkPackages(@NonNull final Properties ctx, @NonNull final List<IQueueProcessor> queueProcessors)
+	{
+		final IQuery<I_C_Queue_WorkPackage> queueProcessorWPQueriesAggregator = queryBL.createQueryBuilder(I_C_Queue_WorkPackage.class)
+				//dev-note: workaround to be able to union multiple queries without applying 'limit' and 'order by'  to the end result
+				.addEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_C_Queue_WorkPackage_ID, -1)
+				.create();
+
+		final int numberOfWorkPackagesPerProcessor = 1;
+
+		for (final IQueueProcessor queueProcessor : queueProcessors)
+		{
+			final IQuery<I_C_Queue_WorkPackage> workPackageQuery = queueProcessor.getQueue().createQuery(ctx, numberOfWorkPackagesPerProcessor);
+
+			final IQuery<I_C_Queue_WorkPackage> nonLockedWorkPackageQuery = lockManager.addNotLockedClause(workPackageQuery);
+
+			queueProcessorWPQueriesAggregator.addUnion(nonLockedWorkPackageQuery, true);
+		}
+
+		return lockManager.retrieveAndLockMultipleRecords(queueProcessorWPQueriesAggregator, I_C_Queue_WorkPackage.class);
+	}
+
+	@NonNull
+	private Optional<IQueueProcessor> getAvailableQueueProcessorForWorkPackage(@NonNull final I_C_Queue_WorkPackage workPackage)
+	{
+		final QueuePackageProcessorId packageProcessorId = QueuePackageProcessorId.ofRepoId(workPackage.getC_Queue_PackageProcessor_ID());
+
+		return queueProcessors
+				.values()
+				.stream()
+				.filter(processor -> processor.getAssignedPackageProcessorIds().contains(packageProcessorId))
+				.filter(IQueueProcessor::isAvailableToWork)
+				.findFirst();
+	}
+
+	@NonNull
+	private Optional<I_C_Queue_WorkPackage> getWorkPackageWithOwnContext(@NonNull final IQueueProcessor queueProcessor, @NonNull final I_C_Queue_WorkPackage workPackage, @NonNull final Properties commonCtx)
+	{
+		final Properties newCtx = Env.copyCtx(commonCtx);
+
+		final Optional<I_C_Queue_WorkPackage> workPackageWithOwnCtx = queryBL
+				.createQueryBuilder(I_C_Queue_WorkPackage.class, newCtx)
+				.addEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_C_Queue_WorkPackage_ID, workPackage.getC_Queue_WorkPackage_ID())
+				.create()
+				.firstOnlyOptional(I_C_Queue_WorkPackage.class);
```

Comment:
hmmm,

2 questions popped in my mind when reading this code:
1. why load it again if we already have it loaded?
2. why it might be not there, the second time when we load it?

Loud thinking, 
I would suggest considering to simply change the context of the current loaded object.
We already have org.compiere.model.PO#setCtx.
So why not introducing InterfaceWrapperHelper.setCtx?

Reply:
well, I don't have a good reason for it... I did so simply because I didn't find another quick way to do it at that time.....